### ISSUE TITLE
Crypto.Hash.keccak can't update() after calling digest()

### ISSUE THREAD

**Issue by czepluch** [Legrandin/pycryptodome#4]
According to the the hashlib interface it should be possible to do the following:
```python
import hashlib
s = hashlib.sha256()
s.update("foo")
s.digest()
s.update("bar")
s.digest()
```
This is however not possible with this library, but it would be very nice if it was.
Trying the same with pycryptodome fails like this:
```python
from Crypto.Hash import keccak
kh = keccak.new(digest_bits=256)
kh.update("foo")
kh.update("bar")
kh.digest()
kh.update("baz")
```
Gives this TypeError:
>  File "/home/czepluch/.virtualenvs/tmp-dab852443eeadef/site-packages/pycryptodome-3.3-py2.7-linux-x86_64.egg/Crypto/Hash/keccak.py", line 111, in update
 >   raise TypeError("You can only call 'digest' or 'hexdigest' on this object")
>TypeError: You can only call 'digest' or 'hexdigest' on this object

Shouldn't pycryptodome follow the same interface as hashlib?

**Comment by Legrandin** [comment]
The hashlib API (or PEP247) is not the greatest so there is no guarantee pycryptodome complies to them.

I am happy to hear about protocol/implementations that would benefit from this change: I can't find any but I can certainly imagine cases where this behavior can backfire and hurt security.

**Comment by czepluch** [comment]
Sure, you might be right. Just noticed that it didn't comply and thought I'd let you know.

**Comment by czepluch** [comment]
Sorry to reopen. Was a bit too quick there. 
Let's say I want to be able to update after calling digest(). Is there a way to wrap this? Am updating a lib to be pypy compatible and this lib uses update after calling digest().

**Comment by czepluch** [comment]
Sorry for all the opening and closing of the issue. Thought I'd found a way to wrap the functions to get the desired functionality, but turns out I didn't. 

I would like to get this functionality, since I am updating a lib that used update() after digest(), to support pypy. Maybe there are chances that it might hurt security in some cases, but that should be up to the user of this lib, to take care of?

**Comment by heikoheiko** [comment]
> I am happy to hear about protocol/implementations that would benefit from this change

This is used to update a MAC. 
https://github.com/ethereum/pydevp2p/blob/develop/devp2p/rlpxcipher.py#L330

**Comment by Legrandin** [comment]
> Maybe there are chances that it might hurt security in some cases, but that should be up to the user of this lib, to take care of?

Yes, I agree with that.

There could be a `update_after_digest` boolean parameter to give to `new`: one can then pass `True` and have the old hashlib-like behavior back. By default, the more conservative (and more secure) logic would be enforced.

**Comment by heikoheiko** [comment]
:+1:

**Comment by Legrandin** [comment]
Fixed with version 3.3.1.

One can create a hash object with:

    from Crypto.Hash import keccak
    hash_object = keccak.new(digest_bits=512, update_after_digest=True)

and be able to invoke`update` at any time.

**Comment by czepluch** [comment]
Awesome!
